src/OrchardCore.Modules/OrchardCore.Tenants/ViewModels/TenantViewModel.cs

```diff
@@ -0,0 +1,105 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Localization;
+using OrchardCore.Data;
+using OrchardCore.Environment.Shell;
+
+namespace OrchardCore.Tenants.ViewModels
+{
+    public class TenantViewModel : IValidatableObject
+    {
+        public string Description { get; set; }
+
+        [Required]
+        public string Name { get; set; }
+
+        public string DatabaseProvider { get; set; }
+
+        public string RequestUrlPrefix { get; set; }
+
+        public string RequestUrlHost { get; set; }
+
+        public string ConnectionString { get; set; }
+
+        public string TablePrefix { get; set; }
+
+        public string RecipeName { get; set; }
+
+        public string FeatureProfile { get; set; }
+
+        [BindNever]
+        public bool IsNewTenant { get; set; }
+
+        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
+        {
+            var shellHost = validationContext.GetService<IShellHost>();
+            var featureProfilesService = validationContext.GetService<IFeatureProfilesService>();
+            var databaseProviders = validationContext.GetServices<DatabaseProvider>();
+            var S = validationContext.GetService<IStringLocalizer<TenantViewModel>>();
+
+            var selectedProvider = databaseProviders.FirstOrDefault(x => x.Value == DatabaseProvider);
+
+            if (selectedProvider != null && selectedProvider.HasConnectionString && String.IsNullOrWhiteSpace(ConnectionString))
+            {
+                yield return new ValidationResult(S["The connection string is mandatory for this provider."], new[] { nameof(ConnectionString) });
+            }
+
+            if (String.IsNullOrWhiteSpace(Name))
+            {
+                yield return new ValidationResult(S["The tenant name is mandatory."], new[] { nameof(Name) });
+            }
+
+            if (!String.IsNullOrWhiteSpace(FeatureProfile))
+            {
+                var featureProfiles = featureProfilesService.GetFeatureProfilesAsync().GetAwaiter().GetResult();
+                if (!featureProfiles.ContainsKey(FeatureProfile))
+                {
+                    yield return new ValidationResult(S["The feature profile does not exist."], new[] { nameof(FeatureProfile) });
+                }
+            }
```

Comment:
We could move this to controller to resolve `async` issue

Reply:
The point of this refactoring is not to duplicate the validation logic, which caused a bug in the past (when validation was different in the two places, that you fixed with this refactoring), and which led you to implement a new one (by just adding validation to the MVC controller). So going back to duplication is not a good idea.

Reply:
Seems the best option that we should go for is to add our own validation interface, which quite similar to `IValidatableObject` that support async. Unless someone suggest a better option